src/meta/src/stream/stream_graph.rs

```diff
@@ -362,10 +362,22 @@ impl StreamGraphBuilder {
         upstream_fragment_id: GlobalFragmentId,
         upstream_actor_ids: &[LocalActorId],
         downstream_actor_ids: &[LocalActorId],
-        exchange_operator_id: u64,
-        dispatch_strategy: DispatchStrategy,
-        same_worker_node: bool,
+        edge: StreamFragmentEdge,
     ) {
+        let exchange_operator_id = edge.link_id;
+        let same_worker_node = edge.same_worker_node;
+        let dispatch_strategy = edge.dispatch_strategy.unwrap();
+        let edge_upstream_id = edge.upstream_id;
+        let edge_downstream_id = edge.downstream_id;
+        // We can't use the exchange operator id directly as the dispatch id, because an exchange
+        // could belong to more than one downstream in DAG. We can use the exchange operator
+        // id together with edge upstream and downstream id as an unique id. In this way we
+        // can ensure the dispatchers of `StreamActor` would have different id, even though they
+        // come from the same exchange operator.
+        let dispatch_id = exchange_operator_id
+            + edge_upstream_id as u64 * 1000
+            + edge_downstream_id as u64 * 1000000;
```

Comment:
IIRC, the dispatcher id has to be unique only in the scope of this actor. So can we directly use the downstream fragment id here?

https://github.com/risingwavelabs/risingwave/blob/181680b58fdbbadf49f8c672d3fc8d43fe84d6c0/proto/stream_plan.proto#L528-L531

For example, MV on MV uses downstream actor ID (actually fragment ID should be enough).

https://github.com/risingwavelabs/risingwave/blob/181680b58fdbbadf49f8c672d3fc8d43fe84d6c0/src/meta/src/stream/stream_manager.rs#L253-L255

Reply:
You are right, we can use edge downstream_id as dispatcher id directly.